ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -0,0 +1,221 @@
+                .onFailure(
+                    new OpenSearchStatusException(
+                        "Error on communication with remote model: " + error.getMessage(),
+                        RestStatus.fromCode(statusCode)
+                    )
+                );
+        }
+    }
+
+    private void processResponse(
+        Integer statusCode,
+        String body,
+        Map<String, String> parameters,
+        Map<Integer, ModelTensors> tensorOutputs
+    ) {
+        ModelTensors tensors;
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            tensors = processErrorResponse("null");
+        } else {
+            if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+                log.error("Remote server returned error code: {}", statusCode);
+                tensors = processErrorResponse(body);
+            } else {
+                try {
+                    tensors = processOutput(body, connector, scriptService, parameters);
+                } catch (Exception e) {
+                    log.error("Failed to process response body: {}", body, e);
+                    tensors = processErrorResponse(body);
+                }
+            }
+        }
+        tensors.setStatusCode(statusCode);
+        tensorOutputs.put(countDownLatch.getSequence(), tensors);
+    }
+
+    private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
+        List<ModelTensors> modelTensors = new ArrayList<>();
+        TreeMap<Integer, ModelTensors> sortedMap = new TreeMap<>(tensorOutputs);
+        log.debug("Reordered tensor outputs size is {}", sortedMap.size());
+        if (tensorOutputs.size() == 1) {
+            // batch API case
+            int status = tensorOutputs.get(0).getStatusCode();
+            if (status == HttpStatus.SC_OK) {
+                modelTensors.add(tensorOutputs.get(0));
+                actionListener.onResponse(modelTensors);
+            } else {
+                try {
+                    actionListener
+                        .onFailure(
+                            new OpenSearchStatusException(
+                                AccessController
+                                    .doPrivileged(
+                                        (PrivilegedExceptionAction<String>) () -> GSON
+                                            .toJson(tensorOutputs.get(0).getMlModelTensors().get(0).getDataAsMap())
+                                    ),
+                                RestStatus.fromCode(status)
+                            )
+                        );
+                } catch (PrivilegedActionException e) {
+                    actionListener.onFailure(new OpenSearchStatusException(e.getMessage(), RestStatus.fromCode(statusCode)));
+                }
+            }
+        } else {
+            // non batch API.
+            for (Map.Entry<Integer, ModelTensors> entry : sortedMap.entrySet()) {
```

Comment:
Just curious, why don't we check `HttpStatus` when tensorOutputs.size() > 1 ?

Reply:
Current implementation sends multiple requests to remote model at the same time, some of them can succeed and some of them can fail. I used a simple solution to use 200 as http status always and user can extract the actual result from the response list, e.g. a sample response could be:
```
{
    "inference_results": [
        {
            "output": [
                {
                    "dataAsMap": {
                        "error_message": {"Runtime exception"}
                    }
                }
            ]
        },
        {
            "output": [
                {
                    "name": "sentence_embedding",
                    "data_type": "FLOAT32",
                    "shape": [
                        768
                    ],
                    "data": [
                        0.08624682,
                        0.28323436,
                        -0.061060455,
                        ...
                    ]
                }
            ]
        }
    ]
}
```

Reply:
> Current implementation sends multiple requests to remote model at the same time, some of them can succeed and some of them can fail.

Why send multiple requests at the same time ?

Reply:
> > Current implementation sends multiple requests to remote model at the same time, some of them can succeed and some of them can fail.
> 
> Why send multiple requests at the same time ?

Previously the requests are send one by one which is natural but not efficient with the sync httpclient. Now we changed this to async httpclient and send multiple requests is a natural approach because we don't need to wait for prior responses.

Reply:
> > I used a simple solution to use 200 as http status always
> 
> Which part defines this logic ? The old code has such logic
> 
> ```
>             if (statusCode < 200 || statusCode >= 300) {
>                 throw new OpenSearchStatusException(REMOTE_SERVICE_ERROR + modelResponse, RestStatus.fromCode(statusCode));
>             }
> ```
> 
> Do you keep the same logic in your new code?

This logic changed. E.g. if we have a requests with 10 sub requests to remote model and previously the 10 requests are sent sequentially and anyone fails will cause the whole request fail, for example first 9 are succeed and the last one fails, then the whole request fail.
The current logic changed since the httpclient is async and we don't need to blocking waiting for any response(for efficiency), and we send 10 requests to remote model at the same time and then collect all the responses and process them together. So if 9 requests succeed and one fails, we return success response(9 success, 1 fail) to user instead of all fail.

Reply:
What happens if 9 fails and 1 succeeds?

Reply:
The overall status code is still 200, but the inner status for each request is different, 9 fails have non 200 status_code and 1 succeed has 200.

Reply:
Will neural search face any issue for this change?

Reply:
This indeed changed the behavior in neural search side, I'll change this part to make it bwc with the existing code logic.

Reply:
Have you already made these changes, or do you need to address them in a separate PR? It would be beneficial to provide more details in the description, given the extensive nature of this PR and the potential for breaking changes. To improve understanding and clarity, please consider adding more information to the issue description, including:

The current user experience and any deviations from the previous implementation, if applicable.
The implications for, or interactions with, our upstream processes (such as Airflow or Neural Search).
This additional information will help in assessing the PR more effectively.

Reply:
The changes already made to this PR, so this PR has exactly same behavior with previous code implementation. So this won't change the upstream behavior.